File: typescript/src/transport.ts
/**
 * Transport abstraction for the client.
 *
 * The client is deliberately decoupled from any concrete WebSocket implementation
 * so it can be unit-tested with a mock and run on Node, the browser, or a custom
 * socket. A transport is anything that can send a string frame and surface
 * incoming string frames + lifecycle events.
 */

export type TransportState = 'connecting' | 'open' | 'closing' | 'closed';

/** Minimal injectable transport contract. Mirrors the browser `WebSocket` subset. */
export interface Transport {
    readonly state: TransportState;
    /** Send a serialized frame. Throws / rejects if the transport is not open. */
    send(data: string): void;
    /** Open the connection. Resolves once the transport reaches `open`. */
    connect(): Promise<void>;
    /** Close the connection. */
    close(code?: number, reason?: string): void;
    /** Register a handler for incoming string frames. Returns an unsubscribe fn. */
    onMessage(handler: (data: string) => void): () => void;
    /** Register a handler for transport close. Returns an unsubscribe fn. */
    onClose(handler: (info: { code?: number; reason?: string }) => void): () => void;
    /** Register a handler for transport-level errors. Returns an unsubscribe fn. */
    onError(handler: (err: unknown) => void): () => void;
}

/** The subset of the standard `WebSocket` interface the default transport needs. */
export interface WebSocketLike {
    readonly readyState: number;
    send(data: string): void;
    close(code?: number, reason?: string): void;
    addEventListener(type: 'open', listener: () => void): void;
    addEventListener(type: 'close', listener: (ev: { code?: number; reason?: string }) => void): void;
    addEventListener(type: 'error', listener: (ev: unknown) => void): void;
    addEventListener(type: 'message', listener: (ev: { data: unknown }) => void): void;
}

export type WebSocketFactory = (url: string) => WebSocketLike;

const WS_CONNECTING = 0;
const WS_OPEN = 1;
const WS_CLOSING = 2;

/** Default connect timeout (ms) for the WebSocket transport. */
const DEFAULT_CONNECT_TIMEOUT = 30_000;

/**
 * Default transport backed by a `WebSocket`-like object. By default it uses the
 * global `WebSocket`; pass a `factory` to inject one (e.g. the `ws` package on
 * Node, or a mock in tests).
 */
export class WebSocketTransport implements Transport {
    private socket: WebSocketLike | null = null;
    private readonly url: string;
    private readonly factory: WebSocketFactory;
    private readonly connectTimeout: number;
    private readonly messageHandlers = new Set<(data: string) => void>();
    private readonly closeHandlers = new Set<(info: { code?: number; reason?: string }) => void>();
    private readonly errorHandlers = new Set<(err: unknown) => void>();

    constructor(url: string, factory?: WebSocketFactory, connectTimeout = DEFAULT_CONNECT_TIMEOUT) {
        this.url = url;
        this.connectTimeout = connectTimeout;
        if (factory) {
            this.factory = factory;
        } else {
            const G = globalThis as { WebSocket?: new (url: string) => WebSocketLike };
            if (!G.WebSocket) {
                throw new Error('No global WebSocket available; pass a WebSocketFactory to WebSocketTransport.');
            }
            const Ctor = G.WebSocket;
            this.factory = (u) => new Ctor(u);
        }
    }

    get state(): TransportState {
        if (!this.socket) return 'closed';
        switch (this.socket.readyState) {
            case WS_CONNECTING:
                return 'connecting';
            case WS_OPEN:
                return 'open';
            case WS_CLOSING:
                return 'closing';
            default:
                return 'closed';
        }
    }

    connect(): Promise<void> {
        if (this.socket && this.socket.readyState === WS_OPEN) return Promise.resolve();

        // A prior socket that never reached OPEN (failed/half-open dial, or a closed
        // socket from a previous attempt) would otherwise be orphaned: its listeners
        // stay registered and keep dispatching into the shared handler sets, so a late
        // message/close from the dead socket double-fires. Close it and detach it
        // before dialing a fresh one. (WebSocketLike has no removeEventListener, so we
        // also guard every handler below with an identity check on `this.socket`.)
        if (this.socket && this.socket.readyState !== WS_OPEN) {
            const stale = this.socket;
            this.socket = null;
            try {
                stale.close();
            } catch {
                // ignore — best-effort teardown of a half-open socket
            }
        }

        return new Promise<void>((resolve, reject) => {
            const socket = this.factory(this.url);
            this.socket = socket;

            let settled = false;
            const timer =
                this.connectTimeout > 0
                    ? setTimeout(() => {
                          if (settled) return;
                          settled = true;
                          // Tear down the half-open socket so it can't leak / fire later.
                          if (this.socket === socket) this.socket = null;
                          try {
                              socket.close();
                          } catch {
                              // ignore
                          }
                          reject(new Error(`WebSocket connect to ${this.url} timed out after ${this.connectTimeout}ms`));
                      }, this.connectTimeout)
                    : undefined;

            socket.addEventListener('open', () => {
                // Ignore events from a socket we've already replaced/abandoned.
                if (this.socket !== socket) return;
                if (settled) return;
                settled = true;
                if (timer) clearTimeout(timer);
                resolve();
            });
            socket.addEventListener('error', (ev: unknown) => {
                if (this.socket !== socket) return;
                for (const h of this.errorHandlers) h(ev);
                if (!settled && this.state !== 'open') {
                    settled = true;
                    if (timer) clearTimeout(timer);
                    if (this.socket === socket) this.socket = null;
                    // Release the failed socket so it can't linger / fire later.
                    try {
                        socket.close();
                    } catch {
                        // ignore
                    }
                    reject(ev instanceof Error ? ev : new Error('WebSocket connection error'));
                }
            });
            socket.addEventListener('close', (ev: { code?: number; reason?: string }) => {
                if (this.socket !== socket) return;
                if (timer) clearTimeout(timer);
                for (const h of this.closeHandlers) h({ code: ev.code, reason: ev.reason });
            });
            socket.addEventListener('message', (ev: { data: unknown }) => {
                if (this.socket !== socket) return;
                const data = typeof ev.data === 'string' ? ev.data : String(ev.data);
                for (const h of this.messageHandlers) h(data);
            });
        });
    }

    send(data: string): void {
        if (!this.socket || this.socket.readyState !== WS_OPEN) {
            throw new Error(`Cannot send: transport is "${this.state}"`);
        }
        this.socket.send(data);
    }

    close(code?: number, reason?: string): void {
        this.socket?.close(code, reason);
    }

    onMessage(handler: (data: string) => void): () => void {
        this.messageHandlers.add(handler);
        return () => this.messageHandlers.delete(handler);
    }

    onClose(handler: (info: { code?: number; reason?: string }) => void): () => void {
        this.closeHandlers.add(handler);
        return () => this.closeHandlers.delete(handler);
    }

    onError(handler: (err: unknown) => void): () => void {
        this.errorHandlers.add(handler);
        return () => this.errorHandlers.delete(handler);
    }
}

File: dotnet/src/Transport.cs
// Transport abstraction for the client.
//
// The client is deliberately decoupled from any concrete WebSocket implementation
// so it can be unit-tested with a mock and run against a real socket in production.
// A transport is anything that can send a string frame and surface incoming string
// frames plus lifecycle (close / error) events.

using System.Net.WebSockets;
using System.Text;

namespace SmooAI.SmoothOperatorAgent;

public enum TransportState
{
    Closed,
    Connecting,
    Open,
    Closing,
}

/// <summary>Info surfaced when a transport closes.</summary>
public readonly record struct TransportCloseInfo(int? Code, string? Reason);

/// <summary>
/// Minimal injectable transport contract. Mockable for tests (no live socket needed).
/// </summary>
public interface ITransport
{
    TransportState State { get; }

    /// <summary>Open the connection. Completes once the transport reaches <see cref="TransportState.Open"/>.</summary>
    Task ConnectAsync(CancellationToken cancellationToken = default);

    /// <summary>Send a serialized frame. Throws if the transport is not open.</summary>
    Task SendAsync(string data, CancellationToken cancellationToken = default);

    /// <summary>Close the connection.</summary>
    Task CloseAsync(int code = 1000, string? reason = null, CancellationToken cancellationToken = default);

    /// <summary>Raised for each incoming string frame.</summary>
    event Action<string>? Message;

    /// <summary>Raised when the transport closes.</summary>
    event Action<TransportCloseInfo>? Closed;

    /// <summary>Raised on a transport-level error.</summary>
    event Action<Exception>? Error;
}

/// <summary>
/// Default transport backed by <see cref="ClientWebSocket"/>. Connects to the given
/// URL, pumps incoming text frames on a background receive loop, and raises
/// <see cref="Message"/> / <see cref="Closed"/> / <see cref="Error"/> events.
/// </summary>
public sealed class WebSocketTransport : ITransport, IAsyncDisposable
{
    private readonly Uri _url;
    private readonly Func<ClientWebSocket>? _factory;
    private ClientWebSocket? _socket;
    private CancellationTokenSource? _receiveCts;
    private Task? _receiveLoop;

    public WebSocketTransport(string url, Func<ClientWebSocket>? webSocketFactory = null)
    {
        _url = new Uri(url);
        _factory = webSocketFactory;
    }

    public event Action<string>? Message;
    public event Action<TransportCloseInfo>? Closed;
    public event Action<Exception>? Error;

    public TransportState State => _socket?.State switch
    {
        WebSocketState.Connecting => TransportState.Connecting,
        WebSocketState.Open => TransportState.Open,
        WebSocketState.CloseSent or WebSocketState.CloseReceived => TransportState.Closing,
        _ => TransportState.Closed,
    };

    public async Task ConnectAsync(CancellationToken cancellationToken = default)
    {
        if (_socket is { State: WebSocketState.Open }) return;

        _socket = _factory?.Invoke() ?? new ClientWebSocket();
        await _socket.ConnectAsync(_url, cancellationToken).ConfigureAwait(false);

        _receiveCts = new CancellationTokenSource();
        _receiveLoop = Task.Run(() => ReceiveLoopAsync(_socket, _receiveCts.Token));
    }

    public async Task SendAsync(string data, CancellationToken cancellationToken = default)
    {
        if (_socket is not { State: WebSocketState.Open })
            throw new InvalidOperationException($"Cannot send: transport is \"{State}\".");

        var bytes = Encoding.UTF8.GetBytes(data);
        await _socket.SendAsync(bytes, WebSocketMessageType.Text, endOfMessage: true, cancellationToken)
            .ConfigureAwait(false);
    }

    public async Task CloseAsync(int code = 1000, string? reason = null, CancellationToken cancellationToken = default)
    {
        _receiveCts?.Cancel();
        if (_socket is { State: WebSocketState.Open or WebSocketState.CloseReceived })
        {
            try
            {
                await _socket.CloseAsync((WebSocketCloseStatus)code, reason ?? string.Empty, cancellationToken)
                    .ConfigureAwait(false);
            }
            catch (Exception ex)
            {
                Error?.Invoke(ex);
            }
        }
        Closed?.Invoke(new TransportCloseInfo(code, reason));
    }

    private async Task ReceiveLoopAsync(ClientWebSocket socket, CancellationToken ct)
    {
        var buffer = new byte[8192];
        var message = new MemoryStream();
        try
        {
            while (!ct.IsCancellationRequested && socket.State == WebSocketState.Open)
            {
                WebSocketReceiveResult result;
                message.SetLength(0);
                do
                {
                    result = await socket.ReceiveAsync(buffer, ct).ConfigureAwait(false);
                    if (result.MessageType == WebSocketMessageType.Close)
                    {
                        Closed?.Invoke(new TransportCloseInfo(
                            (int?)result.CloseStatus, result.CloseStatusDescription));
                        return;
                    }
                    message.Write(buffer, 0, result.Count);
                } while (!result.EndOfMessage);

                var text = Encoding.UTF8.GetString(message.GetBuffer(), 0, (int)message.Length);
                Message?.Invoke(text);
            }
        }
        catch (OperationCanceledException)
        {
            // Normal shutdown.
        }
        catch (Exception ex)
        {
            Error?.Invoke(ex);
            Closed?.Invoke(new TransportCloseInfo(null, ex.Message));
        }
    }

    public async ValueTask DisposeAsync()
    {
        _receiveCts?.Cancel();
        if (_receiveLoop is not null)
        {
            try { await _receiveLoop.ConfigureAwait(false); } catch { /* ignore */ }
        }
        _socket?.Dispose();
        _receiveCts?.Dispose();
    }
}
